Add a variety of key definitions surrounding stability (#142)
open-telemetry/opentelemetry-configuration: README.md

A secondary objective is to allow for stable code generation from the JSON schema for language implementations of the [in-memory configuration model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk.md#in-memory-configuration-model). With stability guarantees in mind, maintainers should be able to write code generation logic (or handcraft an in-memory configuration model) without risk that allowed changes to the schema will break stable implementations. **NOTE**: There is _no_ guarantee that the output of off-the-shelf [code generation](#code-generation) tools will be stable when allowed changes are made. Maintainers are responsible for understanding how code generation tools work (and evolve) and intersect with the guarantees made here.

All schema changes are considered through the lens of maintaining these objectives.

> ### Guarantees and allowed changes

Reviewer: This whole section is written as "absolute" terms.

In reality, every statement here can be invalid, when interpreted in a different, "relative" context.

By absolute/relative, I mean this:

The format is not just a contract between two (2) parties.

The format is a contract between three (3) parties, which are:

* [SCHEMA] maintainers of the opentelemetry-configuration repository, when defining a yaml schema
* [SIG] maintainers of a SIG, when parsing configuration from a yaml schema
* [USER] maintainers of a given user configuration files, when writing a yaml schema configuration

I think it will be beneficial to details separately:

* the contract and expectations between [SCHEMA] and [SIG]
* the contract and expectations between [SCHEMA] and [USER]
* the contract and expectations between [USER] and [SIG]

A claim as "Type property names will not change" is not absolute, some type names will change.

For a given schema version number, the type property name will not change even on minor file_format changes, sure, so this is true in the context of [USER] - [SIG], within some constraints on the file_format value.

This does not forbid name changes when evolving the schema definitions:

* new schema can rename properties, affecting the [SCHEMA] - [SIG] part when parsing, for example on a major version change in file_format
* new schema can rename properties, affecting the [SCHEMA] - [USER] part when writing a yaml config, also on major version change in file_format: users __changing__ the file_format tag in their file are expected to adjust to the new format spec.

Author: > The format is a contract between three (3) parties, which are:

I'm not sure I completely understand what you mean by this. I think part of what you're saying is reflected in the [objectives](https://github.com/jack-berg/opentelemetry-configuration/tree/stability-definition?tab=readme-ov-file#objectives) section, which explains that stability guarantees impact both a user writing YAML config (i.e. [SCHEMA] and [USER] in your terms), and implementations interpretting this config and publishing in-memory representations of the data model (i.e. [SCHEMA] and [SIG] in your terms). 

> A claim as "Type property names will not change" is not absolute, some type names will change.

What do you mean by this? What I mean to claim here is that a given property name (let's say the `.tracer_provider` property for example) which is part of a stable type (i.e. no `*/experimental` suffix) will not change in minor or path versions. 

> new schema can rename properties, affecting the [SCHEMA] - [SIG] part when parsing, for example on a major version change in file_format

Yes, all changes are allowed in major version updates. Do we need to explicitly state this? I'd like to follow the lead of `opentelemetry-proto`'s [stability definitions](https://github.com/open-telemetry/opentelemetry-proto?tab=readme-ov-file#stability-definition) and try to keep language as terse as is required by the language. 

Do you think that it would help to explain that these stability requirements apply to minor version updates, and explain that major version updates can include carte blanche breaking changes.

Author: > A major update in the yaml format will have a massive impact:

I agree the impact is huge.

The impact is so huge that it creates a strong incentive to avoid a major version update. And if the rules are fairly (and necessarily) restrictive in terms of what is allowed in a minor version update, then it creates a strong incentive to model types correctly the first time. 

Consider other areas where similar strict restrictions exist like [opentelemetry-proto](https://github.com/open-telemetry/opentelemetry-proto?tab=readme-ov-file#stability-definition), or the [specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md), or language implementations like [opentelemetry-java](https://github.com/open-telemetry/opentelemetry-java/blob/main/VERSIONING.md#compatibility-requirements). 

We have definitions of what is allowable in minor versions to protect users, and to promote careful / thoughtful design. In all these cases, a major version bump is something to strongly avoid. We write down when a major version is required with the hope that we don't have to do it. 

For your examples:

1. The first example is the removal of a known property / type, the `jaeger` exporter. This is considered a breaking change under the rules laid out in this PR. Luckily for us, `jaeger` is already gone, but the same could happen to other exporters, processors, samplers, etc. The fact that removing a type is a breaking change is an incentive to keep it some form (potentially deprecated) until the next major version, which could be indefinitely if there is never a strong enough reason. Its not an incentive to make invasive changes and bump the major version. 
2. The second example is the adjustment of constraints from allowing `[0, 100]` to `[10, 90]`. This is considered a breaking change under the rules laid out in this PR. I think this is correct. In the migration path you lay out, the change in behavior is broken into multiple minor versions, but this doesn't change the fact that a value of `5` was valid in 1.0.0, and became invalid in 1.2.0. No major version change occurred yet the user's input became invalid. I think this is wrong, regardless of whether or not it is broken into steps. For authors of the schema, the fact that making validation more strict requires a major version bump is an incentive to get the validation criteria correct up front. The PR lists all the different types of validation conditions present in JSON schema for completeness, but I really don't see us using those in most cases. `minLength`, `maxLength`, `format`, `multipleOf` and others seem like really niche keywords we really ought to avoid. I do anticipate (from experience so far) we'll use keywords like `pattern`, `propertyNames`, `minProperties`, `maxProperties`, `required`, and `enum` frequently. It seems reasonable to model these correctly up front. And if we're not sure, we can start off stricter and loosen the criteria as needed, since reducing strictness is allowed.

> Item (b) is about stability garantees between [SCHEMA] and [SIG] as I called it, where the SIG implementation is to expect extra processing beyond yaml parsing, mandated by the schema.

On a somewhat unrelated note: Right now the extra processing responsibilities of the SIG implementation are sometimes implied and other times made explicit in comments. This seems flimsy, and error prone as SIG maintainers may miss some subtle change in the expected semantics. To me, this implies we should minimize these extra responsibilities.

  * [propertyNames](https://json-schema.org/understanding-json-schema/reference/object#propertyNames): will not become stricter.
  * [minProperties, maxProperties](https://json-schema.org/understanding-json-schema/reference/object#size): `minProperties` will not increase, `maxProperties` will not decrease.
  * [required](https://json-schema.org/understanding-json-schema/reference/object#required): will not add additional entries.
  * [contains, minContains, maxContains](https://json-schema.org/understanding-json-schema/reference/array#contains): will not become stricter.
  * [minItems, maxItems](https://json-schema.org/understanding-json-schema/reference/array#length): `minItems` will not increase, `maxItems` will not decrease.
  * [uniqueItems](https://json-schema.org/understanding-json-schema/reference/array#uniqueItems): will not go from `true` to `false`.
  * [enum](https://json-schema.org/understanding-json-schema/reference/enum): will not remove entries.
  * [const](https://json-schema.org/understanding-json-schema/reference/const): will not change.

> * No existing type will be deleted.

Reviewer: Assume a format 1.0.0 (stable) that contains the `Jaeger` exporter, which we all know has been deprecated and removed from the spec.

What this says is that the jeager exporter will never be removed from the yaml format in version 1.x.y, as this requires a major format change.

There must be a reasonable way to deprecate and remove features, and I don't think the following is shocking:

- format 1.0.0 does support a jeager exporter
- format 1.1.0 removes support for jaeger, and documents this as an incompatible breaking change

Author: > There must be a reasonable way to deprecate and remove features, and I don't think the following is shocking:

I think there must be a way to deprecate features, but not removal. Deprecated features must linger on (potentially indefinitely) until a major version bump. This shifts the burden of understanding / dealing with schema evolution from users to maintainers, which is where it belongs.

Semantic conventions reached this same conclusion: https://opentelemetry.io/docs/specs/otel/versioning-and-stability/#deprecated

I believe (and hope) language SIGs reach this same conclusion. In java, we maintain code which we marked as stable but have deprecated with the same gaurantees as other stable code - no removal until next major version, which we have no plans of. Here's an example. We accidentally exposed a [protected constructor](https://github.com/open-telemetry/opentelemetry-java/blob/0ef86fae7d1c568c3f2dabe6c7b536c54000a2d3/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/SpanLimits.java#L52-L58), which is not aligned with our code style in the repository. Even though its unlikely anyone was / is using this, we've maintained it for 3 years, and will continue maintaining it.

#### Extension points

The schema contains types which are designed for extension, as indicated by the presence of `"additionalProperties": true`. For example, [component provider](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk.md#register-componentprovider) provides mechanisms for referencing custom SDK extension components like exporters, processors, samplers, etc. The stability guarantees surrounding properties not explicitly defined in this repository is out of scope. Users should consult documentation for the components interpreting these additional properties and decide if their stability guarantees are sufficient for adoption.

### File format

The top level `.file_format` property is special in that it conveys the version of the schema a user's configuration conforms to. Implementations also target a particular version of the schema, which may or may not align with the version specified by the user.

Given the [guarantees and allowed changes](#guarantees-and-allowed-changes), implementations may encounter the following scenarios:

* The `file_format` major version aligns with the implementation major version, AND:

>   * The `file_format` minor version is less than or equal to the implementation minor version: This is idea, with versions maximally aligned. Despite this, an implementation might not support every property and type of its target version.

Reviewer: Typo: This is `ideal` ?

## Stability definition

**NOTICE**: The stability definitions are applicable after this repository publishes a stable release (1.0.0). Currently, this repository is experimental and makes no stability guarantees. 

Stability definition consists of the following sections:

* [Objectives](#objectives): Overview of the motivation behind stability.
* [Guarantees and allowed changes](#guarantees-and-allowed-changes): Specific details on allowed and disallowed changed.
* [Applicability](#applicability): Limits of stability definitions, including experimental features and extension points.
* [File format](#file-format): The `file_format` property and implementation behavior when schema versions are not aligned.

### Objectives

> The primary objective of stability is to protect users from breaking changes. That is, users providing configuration conforming to a particular stable version of the schema should be able to reliably upgrade minor and patch versions without risk that their configuration becomes invalid, or that the interpretation changes.

Reviewer 2: This assumes semver versioning. We should specify this, right?

Author: Yes, good call. I've added this all the way at the top, so that sections can reference `MAJOR`, `MINOR`, `PATCH` where needed.

> A secondary objective is to allow for stable code generation from the JSON schema for language implementations of the [in-memory configuration model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk.md#in-memory-configuration-model). With stability guarantees in mind, maintainers should be able to write code generation logic (or handcraft an in-memory configuration model) without risk that allowed changes to the schema will break stable implementations. **NOTE**: There is _no_ guarantee that the output of off-the-shelf [code generation](#code-generation) tools will be stable when allowed changes are made. Maintainers are responsible for understanding how code generation tools work (and evolve) and intersect with the guarantees made here.

Reviewer 2: This secondary objective seems much like it is already a part of the primary objective. If we do not want to have users upgrade and have their configuration interpretation change we need to ensure the code that is generated, and ultimately creates their SDK, remains consistently stable.

I'm not sure we can make that guarantee though. I seems like it is going to be up to the implementations to ensure this stability when they generate the code.

Is it enough to say we do not plan to break backwards compatibility of the configuration itself and that we will follow the guidelines listed below to ensure implementations do not break. But it is the responsibility of the implementation to comply with their language specific stability requirements?

> A secondary objective is to allow for stable code generation from the JSON schema for language implementations of the [in-memory configuration model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk.md#in-memory-configuration-model). With stability guarantees in mind, maintainers should be able to write code generation logic (or handcraft an in-memory configuration model) without risk that allowed changes to the schema will break stable implementations. **NOTE**: There is _no_ guarantee that the output of off-the-shelf [code generation](#code-generation) tools will be stable when allowed changes are made. Maintainers are responsible for understanding how code generation tools work (and evolve) and intersect with the guarantees made here.

Reviewer 3: ```suggestion
A secondary objective is to allow for stable code generation from the JSON schema for language implementations of the [in-memory configuration model](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/sdk.md#in-memory-configuration-model). With stability guarantees in mind, maintainers should be able to write code generation logic (or handcraft an in-memory configuration model) without risk that allowed changes to the schema will break stable implementations. **NOTE**: There is _no_ guarantee that the output of off-the-shelf [code generation](#code-generation) tools will be stable when allowed changes are made. Maintainers are responsible for understanding any generated code intersects with the guarantees made here.
```

All schema changes are considered through the lens of maintaining these objectives.

### Guarantees and allowed changes

> A type is the JSON schema analog to a [protobuf message](https://protobuf.dev/programming-guides/proto3/#simple). Types have a [`type`](https://json-schema.org/understanding-json-schema/reference/type) of `object`, and use various keywords to describe their properties and conditions which constitute valid data.

Reviewer 3: ```suggestion
A [type](https://json-schema.org/understanding-json-schema/reference/type) is the JSON schema analog to a [protobuf message](https://protobuf.dev/programming-guides/proto3/#simple). Types have a `type` of `object`, and use various keywords to describe their properties and conditions which constitute valid data.
```

Stable types provide the following guarantees. All types except those excluded in [applicability](#applicability) are considered stable after a 1.0.0 release.

* Type property names will not change.
* The `type` of properties will not change, except the allowed addition of `null`.

> * Type [title](https://json-schema.org/understanding-json-schema/reference/annotations) will not change.

Reviewer 3: Is the title important to capture here?

Author: The title may be used by JSON schema code gen tools. Specifying this allows maintainers to make informed decisions about what they can / cannot rely on.

  * [patternProperties](https://json-schema.org/understanding-json-schema/reference/object#patternProperties): will not expand scope to restrict additional properties.
  * [additionalProperties](https://json-schema.org/understanding-json-schema/reference/object#additionalproperties): will not go from `true` to `false`.
  * [propertyNames](https://json-schema.org/understanding-json-schema/reference/object#propertyNames): will not become stricter.
  * [minProperties, maxProperties](https://json-schema.org/understanding-json-schema/reference/object#size): `minProperties` will not increase, `maxProperties` will not decrease.
  * [required](https://json-schema.org/understanding-json-schema/reference/object#required): will not add additional entries.
  * [contains, minContains, maxContains](https://json-schema.org/understanding-json-schema/reference/array#contains): will not become stricter.
  * [minItems, maxItems](https://json-schema.org/understanding-json-schema/reference/array#length): `minItems` will not increase, `maxItems` will not decrease.

>   * [uniqueItems](https://json-schema.org/understanding-json-schema/reference/array#uniqueItems): will not go from `true` to `false`.

Author: Yes, good catch

  * [enum](https://json-schema.org/understanding-json-schema/reference/enum): will not remove entries.
  * [const](https://json-schema.org/understanding-json-schema/reference/const): will not change.
* No existing type will be deleted.
* No type property will be deleted.

The following additive changes are allowed:

* Adding of new properties to existing types.
* Adding new types.
* Changes that make property validation less strict. See above for examples.

> * Removing a property from `required`.

Reviewer 3: Should marking a feature deprecated be captured here? Maybe we can capture this with some form of annotation

Author: JSON schema has a concept called [annotations](https://json-schema.org/understanding-json-schema/reference/annotations), with a specific `deprecated` keyword: 

> JSON Schema includes a few [keywords](https://json-schema.org/learn/glossary#keyword), that aren't strictly used for validation, but are used to describe parts of a schema.

The key bit here is that annotations are part of validation. Though (and maybe this is what you're getting at) they may be used in code generation, similar to the `title` annotation.

I'm happy to include a note about deprecated here. I think we should be free to add / remove deprecated freely. I.e. it would be acceptable to add it, then later remove it.

Author: I ended up adding bullet points for `description` and `deprecated` annotations.

### File format

The top level `.file_format` property is special in that it conveys the version of the schema a user's configuration conforms to. Implementations also target a particular version of the schema, which may or may not align with the version specified by the user.

Given the [guarantees and allowed changes](#guarantees-and-allowed-changes), implementations may encounter the following scenarios:

* The `file_format` major version aligns with the implementation major version, AND:
  * The `file_format` minor version is less than or equal to the implementation minor version: This is ideal, with versions maximally aligned. Despite this, an implementation might not support every property and type of its target version.

>   * The `file_format` minor version is greater than the implementation minor version: The implementation should detect and emit a warning since there may be configuration features the user specifies which the implementation does not understand. However, this is acceptable in many cases, and not terribly different from the ideal path where an implementation also might not support every configuration feature.

Reviewer 4: So, as an example: minLength is 10 in 1.1.0 and minLength is 5 for 1.2.0 (hence relaxing the value). User of **1.2.0** expects 7 to be a valid value, but for 1.1.0 it is not. What do you do here? Generate a warning _and_ accept the value? Or ignore it?

Author: > User of 1.2.0 expects 7 to be a valid value, but for 1.1.0 it is not. 

So the user has specified that they are using 1.2.0 by setting `file_format: 1.2`. The implementation is conforming to 1.1.0. The implementation detects that the user's version is great than the implementation and emits a warning, alerting the user to the fact that a 1.1.0 implementation won't be able to understand some of the changes from a 1.2.0 file format. 

In this case, the implementation rejects the value as its outside of the acceptable range according to the schema its aware of. If implementations stay up to date with `opentelemetry-configuration` releases, it will be much more common for the implementation's version to be greater than the users (which is ideal) than for the user's version to be greater than the implementations.
